File: sdk/typescript/src/verify-vectors.ts
/** Independent TypeScript verifier for CIGAR canonicalization vectors. */

import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";

type CanonicalNode = boolean | bigint | string | Uint8Array | CanonicalNode[] | Map<string, CanonicalNode>;

class CanonicalFailure extends Error {
  readonly code: string;

  constructor(code: string) {
    super(code);
    this.code = code;
  }
}

class JsonParser {
  private position = 0;
  private readonly source: string;

  constructor(source: string) {
    this.source = source;
  }

  parse(): CanonicalNode {
    const value = this.value();
    this.whitespace();
    if (this.position !== this.source.length) throw new CanonicalFailure("invalid_input");
    return value;
  }

  private whitespace(): void {
    while (/[ \t\r\n]/u.test(this.source[this.position] ?? "")) this.position += 1;
  }

  private value(): CanonicalNode {
    this.whitespace();
    const current = this.source[this.position];
    if (current === '"') return this.string();
    if (current === "[") return this.array();
    if (current === "{") return this.object();
    if (this.source.startsWith("true", this.position)) {
      this.position += 4;
      return true;
    }
    if (this.source.startsWith("false", this.position)) {
      this.position += 5;
      return false;
    }
    if (this.source.startsWith("null", this.position)) throw new CanonicalFailure("null_forbidden");
    return this.number();
  }

  private string(): string {
    const start = this.position;
    this.position += 1;
    let escaped = false;
    while (this.position < this.source.length) {
      const current = this.source[this.position];
      this.position += 1;
      if (escaped) {
        escaped = false;
      } else if (current === "\\") {
        escaped = true;
      } else if (current === '"') {
        try {
          const parsed: unknown = JSON.parse(this.source.slice(start, this.position));
          if (typeof parsed !== "string") throw new CanonicalFailure("invalid_input");
          return parsed;
        } catch (error) {
          if (error instanceof CanonicalFailure) throw error;
          throw new CanonicalFailure("invalid_input");
        }
      }
    }
    throw new CanonicalFailure("invalid_input");
  }

  private array(): CanonicalNode[] {
    this.position += 1;
    const result: CanonicalNode[] = [];
    this.whitespace();
    if (this.source[this.position] === "]") {
      this.position += 1;
      return result;
    }
    for (;;) {
      result.push(this.value());
      this.whitespace();
      const current = this.source[this.position];
      this.position += 1;
      if (current === "]") return result;
      if (current !== ",") throw new CanonicalFailure("invalid_input");
    }
  }

  private object(): Map<string, CanonicalNode> {
    this.position += 1;
    const result = new Map<string, CanonicalNode>();
    this.whitespace();
    if (this.source[this.position] === "}") {
      this.position += 1;
      return result;
    }
    for (;;) {
      this.whitespace();
      if (this.source[this.position] !== '"') throw new CanonicalFailure("invalid_input");
      const key = this.string();
      if (result.has(key)) throw new CanonicalFailure("duplicate_key");
      this.whitespace();
      if (this.source[this.position] !== ":") throw new CanonicalFailure("invalid_input");
      this.position += 1;
      result.set(key, this.value());
      this.whitespace();
      const current = this.source[this.position];
      this.position += 1;
      if (current === "}") return result;
      if (current !== ",") throw new CanonicalFailure("invalid_input");
    }
  }

  private number(): bigint {
    const match = /^-?(?:0|[1-9][0-9]*)(?:\.[0-9]+)?(?:[eE][+-]?[0-9]+)?/u.exec(this.source.slice(this.position));
    if (match === null) throw new CanonicalFailure("invalid_input");
    this.position += match[0].length;
    if (/[.eE]/u.test(match[0])) throw new CanonicalFailure("float_forbidden");
    const value = BigInt(match[0]);
    if (value < -(1n << 63n) || value > (1n << 64n) - 1n) throw new CanonicalFailure("float_forbidden");
    return value;
  }
}

function parseStrictJson(source: string): CanonicalNode {
  return new JsonParser(source).parse();
}

function normalizedJson(value: CanonicalNode): string {
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "bigint") return value.toString();
  if (typeof value === "string") return JSON.stringify(value);
  if (value instanceof Uint8Array) throw new CanonicalFailure("bytes_not_json");
  if (Array.isArray(value)) return `[${value.map(normalizedJson).join(",")}]`;
  const encoder = new TextEncoder();
  const entries = [...value.entries()].sort(([first], [second]) =>
    Buffer.compare(encoder.encode(first), encoder.encode(second)),
  );
  return `{${entries.map(([key, child]) => `${JSON.stringify(key)}:${normalizedJson(child)}`).join(",")}}`;
}

function head(major: number, argument: bigint): Uint8Array {
  const prefix = major << 5;
  if (argument < 24n) return Uint8Array.of(prefix | Number(argument));
  const sizes: Array<[bigint, number, number]> = [
    [0xffn, 24, 1], [0xffffn, 25, 2], [0xffffffffn, 26, 4], [0xffffffffffffffffn, 27, 8],
  ];
  for (const [maximum, additional, size] of sizes) {
    if (argument <= maximum) {
      const output = new Uint8Array(size + 1);
      output[0] = prefix | additional;
      let remainder = argument;
      for (let index = size; index > 0; index -= 1) {
        output[index] = Number(remainder & 0xffn);
        remainder >>= 8n;
      }
      return output;
    }
  }
  throw new CanonicalFailure("limit_exceeded");
}

function concatenate(parts: Uint8Array[]): Uint8Array {
  const output = new Uint8Array(parts.reduce((total, part) => total + part.length, 0));
  let offset = 0;
  for (const part of parts) {
    output.set(part, offset);
    offset += part.length;
  }
  return output;
}

function deterministicCbor(value: CanonicalNode): Uint8Array {
  if (value === false) return Uint8Array.of(0xf4);
  if (value === true) return Uint8Array.of(0xf5);
  if (typeof value === "bigint") return value >= 0n ? head(0, value) : head(1, -1n - value);
  if (value instanceof Uint8Array) return concatenate([head(2, BigInt(value.length)), value]);
  if (typeof value === "string") {
    const encoded = new TextEncoder().encode(value);
    return concatenate([head(3, BigInt(encoded.length)), encoded]);
  }
  if (Array.isArray(value)) return concatenate([head(4, BigInt(value.length)), ...value.map(deterministicCbor)]);
  const entries = [...value.entries()].map(([key, child]) => ({ key: deterministicCbor(key), child }));
  entries.sort((first, second) => Buffer.compare(first.key, second.key));
  return concatenate([head(5, BigInt(entries.length)), ...entries.flatMap(({ key, child }) => [key, deterministicCbor(child)])]);
}

class CborParser {
  position = 0;
  private readonly source: Uint8Array;

  constructor(source: Uint8Array) {
    this.source = source;
  }

  private exact(length: number): Uint8Array {
    const end = this.position + length;
    if (!Number.isSafeInteger(length) || end > this.source.length) throw new CanonicalFailure("invalid_input");
    const value = this.source.slice(this.position, end);
    this.position = end;
    return value;
  }

  private byte(): number {
    const value = this.exact(1)[0];
    if (value === undefined) throw new CanonicalFailure("invalid_input");
    return value;
  }

  private argument(additional: number): bigint {
    if (additional < 24) return BigInt(additional);
    const sizes = new Map([[24, 1], [25, 2], [26, 4], [27, 8]]);
    const size = sizes.get(additional);
    if (size === undefined) throw new CanonicalFailure("non_canonical");
    let value = 0n;
    for (const current of this.exact(size)) value = (value << 8n) | BigInt(current);
    const minimum = new Map([[1, 24n], [2, 0x100n], [4, 0x10000n], [8, 0x100000000n]]).get(size);
    if (minimum === undefined || value < minimum) throw new CanonicalFailure("non_canonical");
    return value;
  }

  parse(): CanonicalNode {
    const initial = this.byte();
    const major = initial >> 5;
    const additional = initial & 31;
    if (major === 0) return this.argument(additional);
    if (major === 1) {
      const value = -1n - this.argument(additional);
      if (value < -(1n << 63n)) throw new CanonicalFailure("limit_exceeded");
      return value;
    }
    if (major === 2 || major === 3) {
      const length = Number(this.argument(additional));
      const data = this.exact(length);
      if (major === 2) return data;
      try {
        return new TextDecoder("utf-8", { fatal: true }).decode(data);
      } catch {
        throw new CanonicalFailure("invalid_input");
      }
    }
    if (major === 4) {
      const length = Number(this.argument(additional));
      return Array.from({ length }, () => this.parse());
    }
    if (major === 5) {
      const length = Number(this.argument(additional));
      const result = new Map<string, CanonicalNode>();
      let previous: Uint8Array | undefined;
      for (let index = 0; index < length; index += 1) {
        const start = this.position;
        const key = this.parse();
        const encoded = this.source.slice(start, this.position);
        if (typeof key !== "string" || (previous !== undefined && Buffer.compare(previous, encoded) >= 0)) {
          throw new CanonicalFailure("non_canonical");
        }
        previous = encoded;
        if (result.has(key)) throw new CanonicalFailure("duplicate_key");
        result.set(key, this.parse());
      }
      return result;
    }
    if (major === 6) throw new CanonicalFailure("non_canonical");
    if (major === 7 && additional === 20) return false;
    if (major === 7 && additional === 21) return true;
    if (major === 7 && additional === 22) throw new CanonicalFailure("null_forbidden");
    if (major === 7 && additional >= 25 && additional <= 27) throw new CanonicalFailure("float_forbidden");
    throw new CanonicalFailure("non_canonical");
  }
}

function strictCbor(source: Uint8Array): CanonicalNode {
  const parser = new CborParser(source);
  const value = parser.parse();
  if (parser.position !== source.length || Buffer.compare(deterministicCbor(value), source) !== 0) {
    throw new CanonicalFailure("non_canonical");
  }
  return value;
}

const domains = new Map([
  ["atom", "CIGAR-ATOM"], ["bundle", "CIGAR-BUNDLE"], ["manifest", "CIGAR-MANIFEST"],
  ["handoff", "CIGAR-HANDOFF"], ["effect", "CIGAR-EFFECT"], ["receipt", "CIGAR-RECEIPT"],
  ["extension_manifest", "CIGAR-EXTENSION-MANIFEST"],
]);

function digest(domain: string, cbor: Uint8Array): Buffer {
  const separator = domains.get(domain);
  if (separator === undefined) throw new Error(`unknown digest domain ${domain}`);
  return createHash("sha256").update(separator).update(Uint8Array.of(0)).update("v1").update(Uint8Array.of(0)).update(cbor).digest();
}

interface ValidVector {
  id: string; domain: string; normalization: string; json_input: string; normalized_json: string; cbor_hex: string;
  digest_hex: string; multihash: string; signature_input_hex: string;
}
interface InvalidVector { id: string; encoding: string; input: string; error: string }
interface DifferentialVector { algorithm: string; count: number; domain: string; digest_accumulator_hex: string }
interface Manifest {
  schema_version: number; profile: string; valid_count: number; invalid_count: number;
  valid: ValidVector[]; invalid: InvalidVector[]; differential: DifferentialVector;
}

function check(condition: boolean, message: string): void {
  if (!condition) throw new Error(message);
}

function differentialRecord(index: number): Map<string, CanonicalNode> {
  const entries: Array<[string, CanonicalNode]> = [
    ["active", index % 2 === 0], ["index", BigInt(index)], ["label", `record-${index % 997}`],
    ["values", [BigInt(index % 17), BigInt(-(index % 19) - 1)]],
  ];
  return new Map(entries);
}

function verify(path: string): [number, number] {
  const manifest = JSON.parse(readFileSync(path, "utf8")) as Manifest;
  check(manifest.schema_version === 1 && manifest.profile === "cigar-canonical-v1", "invalid manifest profile");
  check(manifest.valid_count === manifest.valid.length && manifest.invalid_count === manifest.invalid.length && manifest.valid.length >= 200, "invalid manifest counts");
  for (const vector of manifest.valid) {
    const value = parseStrictJson(vector.json_input);
    if (vector.normalization === "nfc:/human_text") {
      if (!(value instanceof Map) || typeof value.get("human_text") !== "string") throw new Error(`${vector.id}: invalid NFC target`);
      value.set("human_text", (value.get("human_text") as string).normalize("NFC"));
    } else if (vector.normalization !== "none") {
      throw new Error(`${vector.id}: unknown normalization profile`);
    }
    const cbor = deterministicCbor(value);
    const expectedDigest = digest(vector.domain, cbor);
    check(normalizedJson(value) === vector.normalized_json, `${vector.id}: normalized JSON mismatch`);
    check(Buffer.from(cbor).toString("hex") === vector.cbor_hex, `${vector.id}: CBOR mismatch`);
    strictCbor(cbor);
    check(expectedDigest.toString("hex") === vector.digest_hex && `1220${expectedDigest.toString("hex")}` === vector.multihash, `${vector.id}: digest mismatch`);
    check(Buffer.concat([Buffer.from("CIGAR-SIGNATURE\0v1\0"), cbor]).toString("hex") === vector.signature_input_hex, `${vector.id}: signature input mismatch`);
  }
  for (const vector of manifest.invalid) {
    let actual: string | undefined;
    try {
      if (vector.encoding === "json") parseStrictJson(vector.input);
      else if (vector.encoding === "cbor_hex") strictCbor(Buffer.from(vector.input, "hex"));
      else if (vector.encoding === "semantic") throw new CanonicalFailure("invalid_argument");
      else if (vector.encoding === "signature_hex" && Buffer.from(vector.input, "hex").length !== 64) throw new CanonicalFailure("invalid_argument");
    } catch (error) {
      if (error instanceof CanonicalFailure) actual = error.code;
      else throw error;
    }
    check(actual === vector.error, `${vector.id}: expected ${vector.error}, found ${actual ?? "accepted"}`);
  }
  const differential = manifest.differential;
  check(differential.algorithm === "cigar-differential-record-v1" && differential.count >= 100_000, "invalid differential metadata");
  const accumulator = createHash("sha256");
  for (let index = 0; index < differential.count; index += 1) accumulator.update(digest(differential.domain, deterministicCbor(differentialRecord(index))));
  check(accumulator.digest("hex") === differential.digest_accumulator_hex, "100,000-record differential accumulator mismatch");
  return [manifest.valid.length + manifest.invalid.length, differential.count];
}

const path = process.argv[2] ?? "schemas/vectors/canonical-v1.json";
const [vectorCount, differentialCount] = verify(path);
console.log(`verified ${vectorCount} canonical vectors and ${differentialCount} differential records`);

File: sdk/python/src/cigar_sdk/verify_vectors.py
"""Independent Python verifier for CIGAR canonicalization vectors."""

from __future__ import annotations

import hashlib
import json
import sys
from pathlib import Path
from typing import Any


class CanonicalFailure(ValueError):
    """Stable content-free canonicalization failure."""

    def __init__(self, code: str) -> None:
        super().__init__(code)
        self.code = code


def _integer(value: str) -> int:
    parsed = int(value)
    if parsed < -(1 << 63) or parsed > (1 << 64) - 1:
        raise CanonicalFailure("float_forbidden")
    return parsed


def _floating(_value: str) -> float:
    raise CanonicalFailure("float_forbidden")


def _object(pairs: list[tuple[str, Any]]) -> dict[str, Any]:
    result: dict[str, Any] = {}
    for key, value in pairs:
        if key in result:
            raise CanonicalFailure("duplicate_key")
        result[key] = value
    return result


def parse_strict_json(source: str) -> Any:
    """Parse JSON while rejecting duplicates, floats, null, overflow, and trailing data."""

    try:
        value = json.loads(
            source,
            parse_int=_integer,
            parse_float=_floating,
            parse_constant=_floating,
            object_pairs_hook=_object,
        )
    except CanonicalFailure:
        raise
    except (json.JSONDecodeError, UnicodeError) as error:
        raise CanonicalFailure("invalid_input") from error
    _reject_null(value)
    return value


def _reject_null(value: Any) -> None:
    if value is None:
        raise CanonicalFailure("null_forbidden")
    if isinstance(value, list):
        for child in value:
            _reject_null(child)
    elif isinstance(value, dict):
        for child in value.values():
            _reject_null(child)


def normalized_json(value: Any) -> bytes:
    """Render compact UTF-8 JSON with lexicographically sorted keys."""

    return json.dumps(
        value,
        ensure_ascii=False,
        allow_nan=False,
        separators=(",", ":"),
        sort_keys=True,
    ).encode()


def _head(major: int, argument: int) -> bytes:
    prefix = major << 5
    if argument < 24:
        return bytes([prefix | argument])
    if argument <= 0xFF:
        return bytes([prefix | 24, argument])
    if argument <= 0xFFFF:
        return bytes([prefix | 25]) + argument.to_bytes(2, "big")
    if argument <= 0xFFFFFFFF:
        return bytes([prefix | 26]) + argument.to_bytes(4, "big")
    if argument <= 0xFFFFFFFFFFFFFFFF:
        return bytes([prefix | 27]) + argument.to_bytes(8, "big")
    raise CanonicalFailure("limit_exceeded")


def deterministic_cbor(value: Any) -> bytes:
    """Encode the deterministic CIGAR RFC 8949 subset."""

    if value is False:
        return b"\xf4"
    if value is True:
        return b"\xf5"
    if isinstance(value, int):
        if value >= 0:
            return _head(0, value)
        return _head(1, -1 - value)
    if isinstance(value, bytes):
        return _head(2, len(value)) + value
    if isinstance(value, str):
        encoded = value.encode()
        return _head(3, len(encoded)) + encoded
    if isinstance(value, list):
        return _head(4, len(value)) + b"".join(deterministic_cbor(child) for child in value)
    if isinstance(value, dict):
        entries = [(deterministic_cbor(key), child) for key, child in value.items()]
        entries.sort(key=lambda entry: entry[0])
        return _head(5, len(entries)) + b"".join(
            key + deterministic_cbor(child) for key, child in entries
        )
    if value is None:
        raise CanonicalFailure("null_forbidden")
    raise CanonicalFailure("float_forbidden")


class _CborParser:
    def __init__(self, source: bytes) -> None:
        self.source = source
        self.position = 0

    def byte(self) -> int:
        if self.position >= len(self.source):
            raise CanonicalFailure("invalid_input")
        value = self.source[self.position]
        self.position += 1
        return value

    def exact(self, length: int) -> bytes:
        end = self.position + length
        if end > len(self.source):
            raise CanonicalFailure("invalid_input")
        value = self.source[self.position : end]
        self.position = end
        return value

    def argument(self, additional: int) -> int:
        if additional < 24:
            return additional
        sizes = {24: 1, 25: 2, 26: 4, 27: 8}
        size = sizes.get(additional)
        if size is None:
            raise CanonicalFailure("non_canonical")
        value = int.from_bytes(self.exact(size), "big")
        minimum = {1: 24, 2: 0x100, 4: 0x10000, 8: 0x100000000}[size]
        if value < minimum:
            raise CanonicalFailure("non_canonical")
        return value

    def parse(self) -> Any:
        initial = self.byte()
        major, additional = initial >> 5, initial & 31
        if major == 0:
            return self.argument(additional)
        if major == 1:
            value = -1 - self.argument(additional)
            if value < -(1 << 63):
                raise CanonicalFailure("limit_exceeded")
            return value
        if major in (2, 3):
            data = self.exact(self.argument(additional))
            if major == 2:
                return data
            try:
                return data.decode()
            except UnicodeDecodeError as error:
                raise CanonicalFailure("invalid_input") from error
        if major == 4:
            return [self.parse() for _ in range(self.argument(additional))]
        if major == 5:
            result: dict[str, Any] = {}
            previous: bytes | None = None
            for _ in range(self.argument(additional)):
                start = self.position
                key = self.parse()
                encoded = self.source[start : self.position]
                if not isinstance(key, str) or (previous is not None and previous >= encoded):
                    raise CanonicalFailure("non_canonical")
                previous = encoded
                if key in result:
                    raise CanonicalFailure("duplicate_key")
                result[key] = self.parse()
            return result
        if major == 6:
            raise CanonicalFailure("non_canonical")
        if major == 7 and additional in (20, 21):
            return additional == 21
        if major == 7 and additional == 22:
            raise CanonicalFailure("null_forbidden")
        if major == 7 and additional in (25, 26, 27):
            raise CanonicalFailure("float_forbidden")
        raise CanonicalFailure("non_canonical")


def strict_cbor(source: bytes) -> Any:
    """Decode CBOR only if it is already deterministic."""

    parser = _CborParser(source)
    value = parser.parse()
    if parser.position != len(source) or deterministic_cbor(value) != source:
        raise CanonicalFailure("non_canonical")
    return value


DOMAINS = {
    "atom": b"CIGAR-ATOM",
    "bundle": b"CIGAR-BUNDLE",
    "manifest": b"CIGAR-MANIFEST",
    "handoff": b"CIGAR-HANDOFF",
    "effect": b"CIGAR-EFFECT",
    "receipt": b"CIGAR-RECEIPT",
    "extension_manifest": b"CIGAR-EXTENSION-MANIFEST",
}


def digest(domain: str, cbor: bytes) -> bytes:
    """Compute one domain-separated v1 digest."""

    return hashlib.sha256(DOMAINS[domain] + b"\0v1\0" + cbor).digest()


def _differential_record(index: int) -> dict[str, Any]:
    return {
        "active": index % 2 == 0,
        "index": index,
        "label": f"record-{index % 997}",
        "values": [index % 17, -(index % 19) - 1],
    }


def verify(path: Path) -> tuple[int, int]:
    """Verify all golden vectors and the 100,000-record differential gate."""

    manifest = json.loads(path.read_text())
    valid = manifest["valid"]
    invalid = manifest["invalid"]
    if (
        manifest["schema_version"] != 1
        or manifest["profile"] != "cigar-canonical-v1"
        or manifest["valid_count"] != len(valid)
        or manifest["invalid_count"] != len(invalid)
        or len(valid) < 200
    ):
        raise ValueError("invalid vector manifest metadata")
    for vector in valid:
        value = parse_strict_json(vector["json_input"])
        if vector["normalization"] == "nfc:/human_text":
            import unicodedata

            value["human_text"] = unicodedata.normalize("NFC", value["human_text"])
        elif vector["normalization"] != "none":
            raise ValueError(f"unknown normalization profile {vector['normalization']}")
        cbor = deterministic_cbor(value)
        expected_digest = digest(vector["domain"], cbor)
        assert normalized_json(value).decode() == vector["normalized_json"], vector["id"]
        assert cbor.hex() == vector["cbor_hex"], vector["id"]
        assert strict_cbor(cbor) == value, vector["id"]
        assert expected_digest.hex() == vector["digest_hex"], vector["id"]
        assert "1220" + expected_digest.hex() == vector["multihash"], vector["id"]
        assert (b"CIGAR-SIGNATURE\0v1\0" + cbor).hex() == vector["signature_input_hex"], vector["id"]
    for vector in invalid:
        try:
            if vector["encoding"] == "json":
                parse_strict_json(vector["input"])
            elif vector["encoding"] == "cbor_hex":
                strict_cbor(bytes.fromhex(vector["input"]))
            elif vector["encoding"] == "signature_hex" and len(bytes.fromhex(vector["input"])) != 64:
                raise CanonicalFailure("invalid_argument")
            elif vector["encoding"] == "semantic":
                raise CanonicalFailure("invalid_argument")
            else:
                raise ValueError(f"unsupported invalid vector {vector['id']}")
        except CanonicalFailure as error:
            assert error.code == vector["error"], vector["id"]
        else:
            raise AssertionError(f"invalid vector accepted: {vector['id']}")
    differential = manifest["differential"]
    if differential["algorithm"] != "cigar-differential-record-v1" or differential["count"] < 100_000:
        raise ValueError("invalid differential gate metadata")
    accumulator = hashlib.sha256()
    for index in range(differential["count"]):
        accumulator.update(digest(differential["domain"], deterministic_cbor(_differential_record(index))))
    assert accumulator.hexdigest() == differential["digest_accumulator_hex"]
    return len(valid) + len(invalid), differential["count"]


def main() -> None:
    """CLI entry point."""

    path = Path(sys.argv[1] if len(sys.argv) > 1 else "schemas/vectors/canonical-v1.json")
    vector_count, differential_count = verify(path)
    print(f"verified {vector_count} canonical vectors and {differential_count} differential records")


if __name__ == "__main__":
    main()
